Author: Arturo Borrero Gonzalez <arturo.borrero.glez@gmail.com>
Description: Implement the missing distclean option
--- pkg-nftables.orig/Makefile.rules.in	2013-12-22 15:25:02.296809032 +0100
+++ pkg-nftables/Makefile.rules.in	2013-12-22 15:26:12.545157359 +0100
@@ -81,6 +81,7 @@
 .PHONY:			all clean install
 all:			$(SUBDIRS) $(all_targets)
 clean:			$(SUBDIRS) $(clean_targets)
+distclean:		clean
 install:		all $(SUBDIRS) $(install_targets)
 
 .PHONY: $(SUBDIRS)
